client/asset/dcr/coin_selection.go

```diff
@@ -71,69 +72,62 @@ func sumUTXOs(set []*compositeUTXO) (tot uint64) {
 	return tot
 }
 
-// In the following utxo selection functions, the compositeUTXO slice MUST be
-// sorted in ascending order (smallest first, largest last).
-
-// subsetLargeBias begins by summing from the largest UTXO down until the sum is
-// just below the requested amount. Then it picks the (one) final UTXO that
-// reaches the amount with the least excess. Each utxo in the input must be
-// smaller than amt - use via leastOverFund only!
-func subsetLargeBias(amt uint64, utxos []*compositeUTXO) []*compositeUTXO {
-	// Add from largest down until sum is *just under*. Resulting sum will be
-	// less, and index will be the next (smaller) element that would hit amt.
-	var sum uint64
-	var i int
-	for i = len(utxos) - 1; i >= 0; i-- {
-		this := toAtoms(utxos[i].rpc.Amount) // must not be >= amt
-		if sum+this >= amt {
-			break
+// subsetWithLeastSumGreaterThan attempts to select the subset of UTXOs with
+// the smallest total value greater than amt. It does this by making
+// 1000 random selections and returning the best one. Each selection
+// involves two passes over the UTXOs. The first pass randomly selects
+// each UTXO with 50% probability. Then, the second pass selects any
+// unused UTXOs until the total value is greater than or equal to amt.
+func subsetWithLeastSumGreaterThan(amt uint64, utxos []*compositeUTXO) []*compositeUTXO {
+	best := uint64(1 << 62)
+	var bestIncluded *[]bool
+	bestNumIncluded := 0
+
+	iterations := 1000
+	for nRep := 0; nRep < iterations; nRep++ {
+		included := make([]bool, len(utxos))
```

Comment:
This is the main bottleneck, causing ~2000 allocs per call.  It may seem silly, but if we only allocate once and instead zero out the slice after each iteration, it screams.  Copying memory is much faster than allocating memory.


```diff
diff --git a/client/asset/dcr/coin_selection.go b/client/asset/dcr/coin_selection.go
index af90c5a79..2be8a5dbc 100644
--- a/client/asset/dcr/coin_selection.go
+++ b/client/asset/dcr/coin_selection.go
@@ -80,21 +80,24 @@ func sumUTXOs(set []*compositeUTXO) (tot uint64) {
 // unused UTXOs until the total value is greater than or equal to amt.
 func subsetWithLeastSumGreaterThan(amt uint64, utxos []*compositeUTXO) []*compositeUTXO {
        best := uint64(1 << 62)
-       var bestIncluded *[]bool
+       var bestIncluded []bool
        bestNumIncluded := 0
 
-       iterations := 1000
+       rnd := rand.New(rand.NewSource(rand.Int63()))
+
+       included := make([]bool, len(utxos))
+
+       const iterations = 1000
        for nRep := 0; nRep < iterations; nRep++ {
-               included := make([]bool, len(utxos))
 
-               var found bool
                var nTotal uint64
                var numIncluded int
-               for nPass := 0; nPass < 2 && !found; nPass++ {
+       passes:
+               for nPass := 0; nPass < 2; nPass++ {
                        for i := 0; i < len(utxos); i++ {
                                var use bool
                                if nPass == 0 {
-                                       use = rand.Uint32()&1 == 1
+                                       use = rnd.Int63()&1 == 1
                                } else {
                                        use = !included[i]
                                }
@@ -105,15 +108,21 @@ func subsetWithLeastSumGreaterThan(amt uint64, utxos []*compositeUTXO) []*compos
                                        if nTotal >= amt {
                                                if nTotal < best || (nTotal == best && numIncluded < bestNumIncluded) {
                                                        best = nTotal
-                                                       bestIncluded = &included
+                                                       if bestIncluded == nil {
+                                                               bestIncluded = make([]bool, len(utxos))
+                                                       }
+                                                       copy(bestIncluded, included)
                                                        bestNumIncluded = numIncluded
-                                                       found = true
+                                                       break passes // next iter
                                                }
-                                               break
+                                               break // next pass
                                        }
                                }
                        }
                }
+               for i := range included {
+                       included[i] = false
+               }
        }
 
        if bestIncluded == nil {
@@ -121,7 +130,7 @@ func subsetWithLeastSumGreaterThan(amt uint64, utxos []*compositeUTXO) []*compos
        }
 
        set := make([]*compositeUTXO, 0, len(utxos))
-       for i, inc := range *bestIncluded {
+       for i, inc := range bestIncluded {
                if inc {
                        set = append(set, utxos[i])
                }
```


bench:

```go
func Benchmark_subsetFund(b *testing.B) {
	amt := uint64(10e8)
	s := make([]*compositeUTXO, 4000)
	newU := func(amt float64) *compositeUTXO {
		return &compositeUTXO{
			rpc: &walletjson.ListUnspentResult{Amount: amt},
		}
	}

	for i := 0; i < b.N; i++ {
		b.StopTimer()

		for i := range s {
			v := rand.Float64() + float64(rand.Int63n(2))
			v = math.Round(v*1e8) / 1e8
			s[i] = newU(v)
		}
		sort.Slice(s, func(i, j int) bool {
			return s[i].rpc.Amount > s[i].rpc.Amount
		})

		b.StartTimer()

		leastOverFund(amt, s)
	}
}
```

before

```
Benchmark_subsetFund
Benchmark_subsetFund-32    	     727	   1637138 ns/op	 4152855 B/op	    2001 allocs/op
```

after

```
Benchmark_subsetFund
Benchmark_subsetFund-32    	    3734	    332022 ns/op	   46345 B/op	       4 allocs/op
```

Reply:
Related, a `clear()` builtin is coming to Go in 1.21.

https://tip.golang.org/ref/spec#Clear
https://go-review.googlesource.com/c/go/+/448076
https://github.com/golang/go/issues/56351
https://github.com/golang/go/commit/99bc53f5e819c2d2d49f2a56c488898085be3982

Reply:
Nice, looks better. I copied everything you had, except I only put one `break passes` in the outer `if` block.

Reply:
> except I only put one `break passes` in the outer `if` block.

That would seem to change the original method then.  I was just eliminating the `found bool`, but I think it behaves differently now.

Reply:
It does seem to match the cpp now however.  Was just a deviation before?

Reply:
The deviation was because I was considering that the slice was sorted in our case. In the cpp their array is not sorted and if they go over the required amount, they remove the element and keep trying. This actually works much better, but is a bit slower.

I've created a commit here that compares the results and shows the runtime of the shuffled one: cc28045c708
I think we should go for the slower but more accurate one.

Reply:
Seems strange to shuffle first, but OK, let's do that and follow the same approach of remove and keep trying if it goes over. This keep trying approach though would seem to make the `best == amt` break of the outer `nRep` loop more important though, no?

Reply:
Hm yeah a `best == amt` break  would definitely not hurt.